plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/components/upgrade-product-image-gallery/index.tsx

```diff
@@ -0,0 +1,73 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useState, useEffect } from '@wordpress/element';
+import { createBlock, BlockInstance } from '@wordpress/blocks';
+import { dispatch, useSelect } from '@wordpress/data';
+import { UpgradeDowngradeNotice } from '@woocommerce/editor-components/upgrade-downgrade-notice';
+import { findBlock } from '@woocommerce/utils';
+
+const upgradeToBlockifiedProductGallery = (
+	productImageGalleryBlockClientId: string
+) => {
+	const newBlock = createBlock( 'woocommerce/product-gallery' );
+
+	dispatch( 'core/block-editor' ).replaceBlock(
+		productImageGalleryBlockClientId,
+		newBlock
+	);
+
+	return true;
+};
+
+export const UpgradeProductImageGallery = () => {
+	const [ productImageGalleryBlock, setProductImageGalleryBlock ] =
+		useState< BlockInstance | null >( null );
+
+	const getBlocks = useSelect( ( select ) => {
+		return select( 'core/block-editor' ).getBlocks;
+	}, [] );
+
+	useEffect( () => {
+		const foundBlock = findBlock( {
+			blocks: getBlocks(),
+			findCondition: ( block ) =>
+				block.name === 'woocommerce/product-image-gallery',
+		} );
+		if ( foundBlock ) {
+			setProductImageGalleryBlock( foundBlock );
+		} else {
+			setProductImageGalleryBlock( null );
+		}
+	}, [ getBlocks, setProductImageGalleryBlock ] );
```

Comment:
How about something like this:


```suggestion
	const productImageGalleryBlock = useSelect(
	( select ) => {
		const blocks = select( 'core/block-editor' ).getBlocks();
		return findBlock({
			blocks,
			findCondition: ( block ) =>
				block.name === 'woocommerce/product-image-gallery',
		}) ?? null;
	},
	[]
);
```

Reply:
Not a strong opinion, so I don't mind changing this, but the reason I did it the way I is was to avoid `getBlocks()` and `findBlock()` being called many times when the component is rendered. I did a quick test loading the _Single Product_ template and upgrading the non-blockified _Add to Cart with Options_ block to the blockified one. With your suggested approach, `findBlock()` was called ~20 times, while with the original one, only 2.

Reply:
Oof! I guess I had taken for granted that the `useSelect` dependency array worked the same as a `useEffect` dependency array, hence only running it once with an empty array. TIL!